controllers/appgroup_reconciler.go

```diff
@@ -36,25 +41,93 @@ data:
 	dummyConfigmapYAMLName = "templates/dummy-configmap.yaml"
 )
 
-func (r *ApplicationGroupReconciler) reconcile(ctx context.Context, l logr.Logger, appGroup *v1alpha1.ApplicationGroup) (bool, error) {
-	l = l.WithValues(v1alpha1.AppGroupNameKey, appGroup.Name)
-	l.V(3).Info("Reconciling ApplicationGroup object")
+func (r *ApplicationGroupReconciler) reconcileCreateOrUpdate(ctx context.Context, instance *v1alpha1.ApplicationGroup, patch client.Patch) error {
+	instance.Progressing()
+	if err := r.Status().Patch(ctx, instance, patch); err != nil {
+		return err
+	}
+
+	r.Log = r.Log.WithValues(v1alpha1.AppGroupNameKey, instance.Name)
+	r.Log.V(3).Info("Reconciling ApplicationGroup object")
 
-	if len(appGroup.Spec.Applications) == 0 {
-		l.Error(ErrInvalidSpec, "ApplicationGroup must list atleast one Application")
+	if len(instance.Spec.Applications) == 0 {
+		r.Log.Error(ErrInvalidSpec, "ApplicationGroup must list atleast one Application")
 		err := fmt.Errorf("application group must list atleast one Application : %w", ErrInvalidSpec)
-		return false, err
+		return err
 	}
 
-	err := r.reconcileApplications(l, appGroup)
-	if err != nil {
-		l.Error(err, "failed to reconcile the applications")
+	if err := r.reconcileApplications(r.Log, instance); err != nil {
+		r.Log.Error(err, "failed to reconcile the applications")
 		err = fmt.Errorf("failed to reconcile the applications : %w", err)
-		return false, err
+		return err
 	}
-
 	// Generate the Workflow object to submit to Argo
-	return r.generateWorkflow(ctx, l, appGroup)
+	if err := r.generateWorkflow(ctx, r.Log, instance); err != nil {
+		r.Log.Error(err, "failed to reconcile ApplicationGroup instance")
+		return err
+	}
+	instance.Status.ObservedGeneration = instance.Generation
+	return nil
+}
+
+func (r *ApplicationGroupReconciler) reconcileRollback(ctx context.Context, instance *v1alpha1.ApplicationGroup, patch client.Patch, err error) (ctrl.Result, error) {
+	// Rollback to previous successful spec since the annotation was set and this is
+	// an UPDATE event
+	if lastSuccessfulSpec := instance.GetLastSuccessful(); lastSuccessfulSpec != nil {
+		// If this is a HelmRelease failure then we must remediate by cleaning up
+		// all the helm releases deployed by the workflow and helm operator
+		if errors.Is(err, ErrHelmReleaseInFailureStatus) {
+			// Delete the HelmRelease(s) - parent and subchart(s)
+			// Lookup charts using the label selector.
+			// Example: chart=kafka-dev,heritage=orkestra,owner=dev, where chart=<top-level-chart>
```

Comment:
```suggestion

```

Reply:
We can drop this IMO

Reply:
Do you want to drop the entire section or just the comment?

Reply:
Comment